File: mobile/lib/domain/tours/tour_line.dart
import 'dart:math' as math;

/// Die Linie ist das Foto — Spiegel von `src/lib/tours/tourLine.ts`.
class TourLinePoint {
  const TourLinePoint(this.x, this.y);
  final double x;
  final double y;
}

class TourLineFit {
  const TourLineFit({required this.points, required this.loop});
  final List<TourLinePoint> points;
  final bool loop;
  TourLinePoint get start => points.first;
  TourLinePoint get end => points.last;
}

const kTourLineSize = 64.0;
const kTourLinePad = 8.0;
const kTourLineMaxPoints = 64;
const kTourLineLoopPx = 3.0;

List<List<double>> downsampleLngLats(
  List<List<double>> coords, [
  int max = kTourLineMaxPoints,
]) {
  if (coords.length <= max) return coords;
  final step = (coords.length - 1) / (max - 1);
  return [
    for (var i = 0; i < max; i++) coords[(i * step).round()],
  ];
}

TourLineFit? fitTourLine(
  List<List<double>> raw, {
  double width = kTourLineSize,
  double height = kTourLineSize,
  double pad = kTourLinePad,
}) {
  final coords = [
    for (final p in raw)
      if (p.length >= 2 && p[0].isFinite && p[1].isFinite) [p[0], p[1]],
  ];
  if (coords.length < 2) return null;
  final sampled = downsampleLngLats(coords);

  var minLng = double.infinity;
  var maxLng = double.negativeInfinity;
  var minLat = double.infinity;
  var maxLat = double.negativeInfinity;
  for (final p in sampled) {
    minLng = math.min(minLng, p[0]);
    maxLng = math.max(maxLng, p[0]);
    minLat = math.min(minLat, p[1]);
    maxLat = math.max(maxLat, p[1]);
  }
  final midLat = (minLat + maxLat) / 2;
  final midLng = (minLng + maxLng) / 2;
  final cosLat = math.cos(midLat * math.pi / 180).abs().clamp(0.05, 1.0);
  final wDeg = math.max((maxLng - minLng) * cosLat, 1e-9);
  final hDeg = math.max(maxLat - minLat, 1e-9);
  final scale = math.min((width - pad * 2) / wDeg, (height - pad * 2) / hDeg);

  final points = [
    for (final p in sampled)
      TourLinePoint(
        width / 2 + (p[0] - midLng) * cosLat * scale,
        height / 2 + (midLat - p[1]) * scale,
      ),
  ];
  final start = points.first;
  final end = points.last;
  final loop = math.sqrt(
        math.pow(end.x - start.x, 2) + math.pow(end.y - start.y, 2),
      ) <
      kTourLineLoopPx;
  return TourLineFit(points: points, loop: loop);
}

List<List<double>> trackCoordsOf({
  required List<List<double>> coordinates,
  List<List<double>> tour = const [],
}) {
  if (coordinates.length >= 2) return coordinates;
  if (tour.length >= 2) return tour;
  return const [];
}

File: src/lib/tours/tourLine.ts
/**
 * Die Linie ist das Foto.
 *
 * Komoot/Strava zeigen ein Hero-Foto oder Kacheln. FlowLine zeigt die echte
 * Spur — orange auf Charcoal, ohne erfundenes Höhenprofil und ohne Stockfoto.
 * Kein Track → null (UI nimmt das gestrichelte No-Track-SVG).
 */

export type LngLat = [number, number];

export type TourLinePoint = { x: number; y: number };

export type TourLineFit = {
  d: string;
  start: TourLinePoint;
  end: TourLinePoint;
  loop: boolean;
  pointCount: number;
};

export const TOUR_LINE_SIZE = 64;
export const TOUR_LINE_PAD = 8;
export const TOUR_LINE_MAX_POINTS = 64;
/** Projected start/end closer than this (in px at SIZE) counts as a loop. */
export const TOUR_LINE_LOOP_PX = 3;

export function downsampleLngLats(
  coords: LngLat[],
  max = TOUR_LINE_MAX_POINTS,
): LngLat[] {
  if (coords.length <= max) return coords;
  const step = (coords.length - 1) / (max - 1);
  const out: LngLat[] = [];
  for (let i = 0; i < max; i++) {
    const p = coords[Math.round(i * step)];
    if (p) out.push([Number(p[0]), Number(p[1])]);
  }
  return out;
}

function asLngLat(raw: ArrayLike<number>): LngLat | null {
  if (raw.length < 2) return null;
  const lng = Number(raw[0]);
  const lat = Number(raw[1]);
  if (!Number.isFinite(lng) || !Number.isFinite(lat)) return null;
  return [lng, lat];
}

export function fitTourLine(
  raw: ReadonlyArray<ArrayLike<number>>,
  width = TOUR_LINE_SIZE,
  height = width,
  pad = TOUR_LINE_PAD,
): TourLineFit | null {
  const coords: LngLat[] = [];
  for (const p of raw) {
    const hit = asLngLat(p);
    if (hit) coords.push(hit);
  }
  if (coords.length < 2) return null;
  const sampled = downsampleLngLats(coords);

  let minLng = Infinity;
  let maxLng = -Infinity;
  let minLat = Infinity;
  let maxLat = -Infinity;
  for (const [lng, lat] of sampled) {
    if (lng < minLng) minLng = lng;
    if (lng > maxLng) maxLng = lng;
    if (lat < minLat) minLat = lat;
    if (lat > maxLat) maxLat = lat;
  }
  const midLat = (minLat + maxLat) / 2;
  const midLng = (minLng + maxLng) / 2;
  const cosLat = Math.min(1, Math.max(0.05, Math.abs(Math.cos((midLat * Math.PI) / 180))));
  const wDeg = Math.max((maxLng - minLng) * cosLat, 1e-9);
  const hDeg = Math.max(maxLat - minLat, 1e-9);
  const scale = Math.min((width - pad * 2) / wDeg, (height - pad * 2) / hDeg);

  const points: TourLinePoint[] = sampled.map(([lng, lat]) => ({
    x: width / 2 + (lng - midLng) * cosLat * scale,
    y: height / 2 + (midLat - lat) * scale,
  }));

  const start = points[0];
  const end = points[points.length - 1];
  if (!start || !end) return null;
  const dx = end.x - start.x;
  const dy = end.y - start.y;
  const loop = Math.hypot(dx, dy) < TOUR_LINE_LOOP_PX;

  const d = points
    .map((p, i) => `${i === 0 ? "M" : "L"}${round(p.x)} ${round(p.y)}`)
    .join(" ");

  return { d, start, end, loop, pointCount: points.length };
}

function round(n: number): string {
  return (Math.round(n * 100) / 100).toFixed(2);
}
